### PR TITLE
perf(pro): optimize native ArrayCache updates and slicing

### PR DESCRIPTION
**Pull request by pcriadoperez**
## Summary

Optimize hand-written ArrayCache internals in Python, PHP, C#, and Go while preserving tested ordering, eviction, merges, row references, conversions, errors, and update counters. Production TS/JS, Java, and Rust cache implementations are unchanged; their tests are expanded.

- Python traverses large contiguous slices once; short/strided slices retain indexed reads. If a slice-bound callback changes the storage length, use the original indexing path and errors.
- PHP retains forward first-match lookup and uses `array_pop` only when the match is last in aligned dense storage; otherwise retain `array_splice`.
- C# avoids temporary key-list allocations for ordinary nonempty string dictionary keys, with the original `SafeString` fallback.
- Go hoists raw, post-merge incoming key reads outside the forward scan.

The original Python/PHP tail-lookup shortcut has been **removed**: broader testing found that public hashmap edits can leave duplicate index tokens, changing first-match behavior. Regression tests now pass on both the pre-PR implementation and revised code. Python append behavior/code is restored; no Python append speedup is claimed.

## Performance

Local Apple M4 Pro microbenchmarks, median microseconds per operation (not end-to-end exchange latency):

| Workload | Rows | Original | Revised |
| --- | ---: | ---: | ---: |
| Python full slice | 1,000 | 17.78 | 5.61 |
| Python full slice | 10,000 | 372.82 | 45.19 |
| PHP newest-order update | 1,000 | 7.64 | 3.91 |
| PHP newest-order update | 10,000 | 74.40 | 34.50 |
| C# newest-order update, Release | 1,000 | 33.72 | 9.60 |
| C# newest-position update, Release | 1,000 | 75.02 | 16.93 |

C# allocations fall from 88,424 to 336 bytes per order update and 176,328 to 328 bytes per position update. Earlier Go measurements were 17.10 → 9.80 us/update; subsequent runs during concurrent builds were noisy, so those should not be treated as stable latency guarantees. Go is reproducible via `go test ./v4 -run '^$' -bench BenchmarkArrayCacheRepeatedTail -benchtime=500ms -count=3` from `go/`.

Older-row updates remain linear. PHP rotating updates were approximately unchanged. Short Python slices have a small dispatch overhead; this does not promise that every workload becomes faster.

## Before/after verification

Baseline is pinned to **63951db9f44dcf425c81785d65867dbb1ac6abf1**, before the PR, not the already-optimized HEAD. Original implementations were loaded separately or substituted with Go overlays; working-tree production code was not swapped during the comparisons.

- Python: identical shared/native suites pass on both revisions. 18,252 slice cases cover the 64/65 boundary, empty/reverse/strided selections, huge bounds/steps and identity. Custom `__index__` mutation retains the original error. A 125,000-operation differential replay passes across all five variants and bounded/unbounded capacities, checking rows and bookkeeping after each operation.
- PHP: identical shared/native suites pass on both revisions, with shared assertions enabled and **222 explicit native assertions** per run. Another 125,000-operation differential replay passes across all five variants. Coverage includes duplicate tokens, reference aliases, sparse/desynchronized arrays, JSON, iteration, polling and clear.
- C#: shared and native behavioral/concurrency suites pass with both original and revised cache classes. Native helper-only probes use the original `SafeString` on the baseline, where the new helper does not exist. Four workers exercise append/poll on each keyed variant; inspect mutable row contents after writers join.
- Go: identical native tests pass on both implementations. `-race` cache suites pass three repeats on each for supported concurrent append/snapshot operations. The reviewer's integer-id coverage gap is fixed; additional raw-key tests cover symbol/outcome and integer/int64/float/bool/string values. A deliberately incorrect normalized-key mutation fails both strengthened tests.
- The duplicate-token and slice-callback regressions were observed failing before the follow-up fixes, and passing on the original and revised implementations.

## Concurrency limitations — not a race-free certification

The tested serialized behavior is equivalent; this PR does **not** establish universal thread safety or fix existing concurrency defects.

- A separate Go `Append`/`GetLimit` probe reports **DATA RACE on both original and revised code**. `GetLimit` accesses counters outside the append mutex. Timestamp polling has the same missing-lock pattern. `Clear` resets storage and metadata in separate critical sections, creating an existing interleaving risk identified by inspection. These need a separate locking fix, not an unrelated change hidden in a performance PR.
- Go/C# snapshots retain shared mutable row maps/dictionaries. Direct external reads/writes during an update are not protected by container locks.
- Python/PHP cache use must be serialized. Python's iterator-based slice traversal can raise `RuntimeError` under concurrent storage mutation; the old indexed loop could instead return mixed state or raise `IndexError`. There is no cross-thread snapshot guarantee.
- Public storage/hashmap/key mutation can already corrupt bookkeeping. The new tests preserve specific existing outcomes; they do not make arbitrary corruption supported.

## Coverage — all ArrayCache variants

Expanded shared tests plus native-language tests cover ordering, eviction, identity, merges, counters, limits, clear, malformed inputs and native fallback paths. These are measured cache-only figures, not whole-library coverage:

| Language | Coverage |
| --- | --- |
| TypeScript-generated JS | 100% lines / branches / functions |
| Python | 243/243 statements; 68/68 branches |
| PHP | 195/195 executable lines; branches not measured |
| C# | 100% lines / branches / methods |
| Go | 281/281 statements, including BaseCache and Set |
| Java | 189/189 lines; 94/94 branches; 32/32 methods; 37 passing JUnit tests |
| Rust factories | 35/35 lines; 6/6 functions |
| Rust Value cache helper ranges | 400/412 executable lines (97.1%) |

Rust's remaining ten helper lines are defensive/impossible-state guards; two more are shared OrderBook cache extraction. They are not excluded to inflate the result. This is not literal 100% Rust runtime coverage or a universal race-free guarantee. Details and runnable native/live test entry points: `build/tests/README.md`.

## Build and offline verification — 2026-09-08

- [x] Full `npm run lint` and `npm run tsBuild` pass. Upstream merge `b35f0a197cf` brings the Paradex lint and Rust test-aggregator fixes; no hook was bypassed.
- [x] Shared cache tests regenerated for Python/PHP, C#, Go and Rust and exercised in their real runners. Only sources/handwritten tests committed; generated drift restored afterward.
- [x] Full standard WebSocket base suites pass in JS, Python, PHP, C#, Go and Rust. C# uses an isolated .NET 10.0.400 SDK. Rust full runner built and passed offline; its lockfile required only four local-package version refreshes, which were not committed.
- [x] Java: all 37 JUnit cache tests pass, including a fresh run after the upstream merge.
- [x] Original-versus-optimized Python/PHP/C#/Go cache suites pass against the pinned pre-performance revision, including the expanded native/shared cases. These are behavioral checks, not proof over arbitrary concurrent mutation.
- [x] Full Python style and PHP syntax checks; new Python harness ruff; Rust native-test formatting; `git diff --check`; real pre-push checks all pass.
- [ ] Full exchange-wide transpile/build/request/response/ID matrices not rerun locally: no exchange implementation changed. Three fresh remote builds have passed and four remain in progress for `b53fad30b5a`; their results are separate from the passing local cache gates.

## Public live shadow — 2026-09-08

Passed `python build/tests/arrayCacheShadow.py --live --seconds 180`, with **zero shadow mismatches**:

| newUpdates | Duration | watchOrderBook results | watchTrades results | watchOHLCV results | Compared book/trade/OHLCV frames | Large-slice checks |
| --- | --- | ---: | ---: | ---: | --- | ---: |
| true | 180 s | 4,296 | 121 | 123 | 6,215 / 131 / 132 | 83 |
| false | 180 s | 4,413 | 72 | 75 | 6,100 / 77 / 78 | 38 |

One public Kraken BTC/USD connection per mode feeds identical frames to current and pinned-baseline Python handlers/cache classes. Compares full order books, cache contents, bookkeeping, slices, and 244 actual update-limit polls in newUpdates mode; validates returned structures. Both modes exercise caches larger than 64 rows. The offline preflight also detects deliberate shadow corruption. The first 45-second attempt failed its minimum large-cache coverage requirement; the longer runs above passed without relaxing that requirement.

No API keys, private endpoints, orders or other exchange mutations. Connections close in finally. This certifies the tested public Python path, not all venues/languages, latency, wire checksums, or thread safety. Other languages are covered by the offline suites above; the fresh broader GitHub live matrix is still running.

## Go locking follow-up

The separate correctness fix remains draft PR #30314, head `3e049574fd3`: shared-mutex polling and atomic clear. Its review requests were addressed with an explicit `go test -C go -race -count=1 -run '^TestArrayCache' ./v4` CI gate and a BaseCache.Clear subclass warning. Local race verification passed 20 repetitions; the expanded cache suite additionally passed three race-enabled overlay repetitions with 100% cache statements. This performance PR does not include that locking change.

Python's serialized/event-loop ownership contract is documented in the now-published `52548c5da64`. Direct concurrent mutations of shared rows remain outside the container-lock guarantees.

### CODE REVIEW

File: go/v4/exchange_cache_test.go [hunk 1]
```diff
@@ -2,6 +2,40 @@ package ccxt
...
+func TestArrayCacheUpdateScanUsesRawKeys(t *testing.T) {
+	cache := NewArrayCacheBySymbolById(3)
+	first := map[string]any{"symbol": "A", "id": 1, "retained": true}
+	cache.Append(first)
+	cache.Append(map[string]any{"symbol": "B", "id": 1})
+	cache.Append(map[string]any{"symbol": "A", "id": "2"})
+	// The bucket normalizes ids to strings, but the row scan compares raw
+	// values after merging. Hoisting must not substitute the normalized id.
+	cache.Append(map[string]any{"symbol": "A", "id": "1", "updated": 1})
+	cache.Append(map[string]any{"symbol": "A", "id": "1", "updated": 2})
```
Comment (carlotestor) [inline go/v4/exchange_cache_test.go]: This scenario can't detect the substitution it guards against. The `"id": "1"` update on line 13 merges into `first` and overwrites its raw `id` with the string `"1"`, so by the time the scan runs the row's raw id and the hashmap's normalized key are both `"1"` — hoisting either one behaves identically.

Confirmed by injecting the exact mistake the comment names: replacing the hoist with `itemId := any(id); itemKey := any(symbol)` still passes this test. Keeping the repeated update's id as the int `1` does fail it (rows come back in order `[1 2]` instead of `[2 1]`).

Suggest making the last two appends use `"id": 1` so the raw-vs-normalized distinction is observable.

**Review by carlotestor — COMMENTED** [review]
## Summary

Checked out `074c331` and built/ran each backend rather than reading the diff. All four native suites are green, and the optimizations are behaviour-neutral where it counts.

| Suite | Result |
| --- | --- |
| `python/ccxt/pro/test/base/test_cache_native.py` | pass |
| `php/pro/test/base/test_cache_native.php` | pass (207 assertions) |
| `go test -run TestArrayCache ./v4` | 4/4 pass |
| C# `dotnet run -- --baseTests --ws` | pass, incl. both new tests |
| shared transpiled `test_cache.{py,php}` | pass (no regression) |

**Equivalence.** Differential fuzz of `ArrayCacheBySymbolById` / `ByOutcomeById` / `BySymbolBySide`, master vs this head: 60,000 appends over randomized max sizes with mixed int/str ids — **0 behavioural differences** in contents, ordering, `get_limit`, or raised exception type. A separate 72,000-append run never produced a duplicate private-index token or an `_index`/`_deque` length desync, which is the invariant the `_index[-1] == token` fast path rests on.

The slice rewrite was fuzzed against Python's own list semantics — 43,200 comparisons (forward/reverse/strided, clipped and out-of-range bounds, `10**18` steps, sizes 0–600): 0 mismatches, element identity preserved.

Perf reproduces on x86-64 Linux (your table is Apple M4 Pro), same order of magnitude:

| Case | master | this head | speedup |
| --- | ---: | ---: | ---: |
| repeated tail update, 1000 rows | 9.80 us | 0.71 us | 13.8x |
| full slice, 1000 rows | 16.62 us | 4.71 us | 3.5x |
| full slice, 10000 rows | 559.53 us | 37.17 us | 15.1x |

### One coverage gap

`TestArrayCacheUpdateScanUsesRawKeys` does not actually pin the behaviour its comment names — detail inline. The Go production change is still correct (`item` is never reassigned inside the scan, so the hoist reads identical values); this is a test-strength issue, not a bug.

The C# fallback test, by contrast, is genuine: dropping the `text.Length > 0` guard in `cacheStringKey` fails it at `ArrayCacheRegressionTest.cs:363`, and it passes again once restored.

### Two notes on the checklist

- **C#** — `cs/tests` targets `net10.0` and builds and runs cleanly here on .NET SDK `10.0.111`: `dotnet build` succeeds, then `dotnet run --no-build -- --baseTests --ws` prints `[C#] ArrayCache regression tests passed`. You may be able to tick that box directly instead of relying on the separate .NET 9 harness.
- **PHP** — `composer.json` requires `php >= 8.1.0` and `array_is_list` is already used throughout `php/Exchange.php`, so the `array_pop` guard is safe on the supported floor.

Merge gate: 🟢
Merge probability: 83%

File: python/ccxt/async_support/base/ws/cache.py [hunk 2]
```diff
@@ -48,8 +49,31 @@ def __getitem__(self, item):
...
         if isinstance(item, slice):
-            start, stop, step = item.indices(len(deque))
-            return [deque[i] for i in range(start, stop, step)]
+            size = len(deque)
+            start, stop, step = item.indices(size)
+            indices = range(start, stop, step)
+            count = len(indices)
+            if count == 0:
+                return []
+            if count <= 64 or abs(step) != 1 or len(deque) != size:
```
Comment (carlotestor) [inline python/ccxt/async_support/base/ws/cache.py:67]: Good guard — bailing to the per-index probe when `__index__` resized the deque keeps the original `IndexError` semantics. Note this also means the fast traversal is skipped for one call after any mutation, which is fine here since the mutation case is pathological.

File: php/pro/ArrayCacheBySymbolById.php [hunk 3]
```diff
@@ -41,10 +41,16 @@ public function append($item) {
             # collides, eg ('BTC/USDT1', '2') and ('BTC/USDT', '12') both
             # yield "BTC/USDT12", whereas "9:BTC/USDT12" and "8:BTC/USDT12"
             # are distinct for every possible pair
+            $last_index = count($this->index) - 1;
             $index = array_search($this->index_key($key, $id), $this->index, true);
             # a miss must not splice - array_splice() coerces false to 0 and
             # would silently remove the first row
-            if ($index !== false) {
+            if ($index !== false && $index === $last_index && count($this->deque) === $last_index + 1 && array_is_list($this->deque)) {
```
Comment (carlotestor) [inline php/pro/ArrayCacheBySymbolById.php:48]: With the `$at_tail` hoist gone, `array_search` is a full scan again and this branch only saves the `array_splice` copy, not the lookup. That is the right correctness trade-off, but the "PHP, by id 7.66 -> 0.46 (16.6x)" row in the PR description was measured on the removed short-circuit — please re-benchmark before undrafting.

**Review by carlotestor — COMMENTED** [review]
Follow-up on `47f99f4` ("fix(pro): preserve ArrayCache edge cases and strengthen equivalence tests") — delta only; nothing from the earlier `074c331` pass repeated.

What this commit does: it **reverts the tail fast-path**. Python `ArrayCacheBySymbolById.append` / `ArrayCacheBySymbolBySide.append` go back to a plain `self._index.index(token)`, and PHP goes back to `array_search($this->index_key(...), $this->index, true)` instead of the hoisted `$token` + `$at_tail` short-circuit — the `array_pop` dense-array shortcut now fires only when the *found* index really is the last one. Python's slice path also gains a `len(deque) != size` guard so a mutating `__index__` bound still raises the original `IndexError`. Test surface grows: Python slice sizes/bounds around the 64-element threshold plus a `ShrinkingBound` case, a PHP duplicate-token first-match test, a Go raw-key-type matrix, and Go/C# concurrent append+snapshot workloads.

The delta looks correct to me. Dropping the `$this->index[$last_index] === $token` shortcut restores first-match semantics when a public `hashmap` edit leaves duplicate tokens, and the new PHP test proves the detached original row is not updated. The slice guard is the right conservative choice — re-deriving `indices` after a callback mutated storage would have silently returned a different window instead of raising.

One thing to fix before this leaves draft: the PR description no longer matches the code. The 18.8x Python and 16.6x PHP "by id" numbers came from exactly the tail short-circuit this commit removes; repeated updates of the newest row are O(retained window) again in both languages. The summary paragraph "Python/PHP check the final private index token before scanning" is now false as well. The remaining wins (Python contiguous slice traversal, C# key reads, Go raw-key hoist) are unaffected. Please re-measure and update the table + prose so reviewers are not comparing against a version that no longer exists.

Minor, non-blocking: the Go concurrency test only checks capacity/counters, while the C# analogue additionally polls `getLimit` from the workers. If that asymmetry is deliberate (Go `GetLimit` not being safe to poll concurrently), a one-line comment saying so would stop a future reader from "fixing" the gap.

Merge gate: 🟢

File: php/pro/ArrayCacheBySymbolById.php (hunk 3 above)
Comment (pcriadoperez) [reply php/pro/ArrayCacheBySymbolById.php:48]: Re-benchmarked after removing the duplicate-token-unsafe tail shortcut. The current PR description reports the revised PHP results: repeated tail updates at 1,000 entries 7.64→3.91 μs/op and at 10,000 entries 74.40→34.50 μs/op (roughly 2×, not the superseded larger claim). The forward first-match search remains; array_pop is used only when that match is the aligned dense tail. Expanded native and shared suites also pass against both the original and optimized PHP implementations.

**Comment by pcriadoperez** [comment]
Coverage follow-up: expanded tests are committed locally as 6275538883d, but publishing is blocked by the required pre-push lint errors in unchanged Paradex lines 1113/1122; the hook was not bypassed. Measured cache coverage is 100% lines/branches/functions in JS, 243/243 statements and 68/68 branches in Python, 195/195 executable lines in PHP (branches not measured), 100% lines/branches/methods in C#, 281/281 statements in Go, and 189/189 lines plus 94/94 branches in Java (37 passing tests). Expanded original/optimized comparisons pass in Python, PHP, C# and Go. Standard WS base runners pass in JS/Python/PHP/Go; C# cache suites pass in a .NET 9 harness but the standard test project requires unavailable .NET 10. Rust shared cache tests pass in isolation, but runtime helper coverage remains incomplete (311/412 selected cache-helper lines, not a full Value dispatch metric); the full regenerated Rust runner hits the existing malformed LBank ping-test identifier. The requested live shadow test has NOT been run because the preceding full gates are not yet green. The earlier live smoke in the PR body is not a substitute for that shadow test. Go locking review fixes and a race CI gate are published separately in #30314 (3e049574fd3), with 20 passing race-suite repeats and three additional passing expanded-suite repeats.

File: build/rustTranspiler.ts [hunk 4]
```diff
@@ -8207,12 +8207,16 @@ impl std::ops::DerefMut for ${coreName} {
             '#![allow(non_snake_case, unused, dead_code, clippy::all)]',
             '',
             modLines,
+            ...(names.includes('test.cache') ? [
+                '#[path = "../../../build/tests/arrayCacheRust.rs"] mod cache_native;',
+            ] : []),
```
Comment (carlotestor) [inline build/rustTranspiler.ts:8212]: The injected `#[path = "../../../build/tests/arrayCacheRust.rs"]` hard-codes the depth from the emitted aggregator to the repo root. `rust/` is gitignored and fully regenerated, so if the aggregator's output directory ever moves one level the module silently fails to compile rather than skipping — and the failure surfaces as an unrelated rustc error. Consider deriving the relative path from the same variable that decides the aggregator's output location instead of a literal.

File: build/tests/arrayCacheShadow.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+from pathlib import Path
+import subprocess
+import sys
+import types
+
+ROOT = Path(__file__).resolve().parents[2]
+sys.path.insert(0, str(ROOT / 'python'))
+import ccxt.pro  # noqa: E402
+
+BASELINE = '63951db9f44dcf425c81785d65867dbb1ac6abf1'
```
Comment (carlotestor) [inline build/tests/arrayCacheShadow.py:21]: `baseline_exchange()` resolves the pinned baseline with `git show 63951db:...`, which needs the commit present locally. A shallow CI checkout (`fetch-depth: 1`, the default for most workflows here) will make this raise `CalledProcessError` at import time — including in the default offline mode. Either guard it with a clear message telling the runner to `git fetch --depth=... 63951db`, or skip the shadow when the object is missing.

**Review by carlotestor — COMMENTED** [review]
Follow-up on `b53fad3` ("test(pro): verify native Rust caches and public live shadows") — delta only; nothing from the earlier `074c331` / `47f99f4` passes repeated.

This commit is **test/harness-only**: no production cache code changed. It adds `build/tests/arrayCacheRust.rs` (262 hand-written Rust cache assertions wired into the generated WS aggregator by `build/rustTranspiler.ts`), `build/tests/arrayCacheShadow.py` (opt-in Kraken shadow replay against the pinned pre-PR baseline `63951db`), and a README describing both.

What I checked in the delta:
- The Rust suite covers the cases the earlier revert was about — shared clones, hashmap write-through with a side-keyed row that has no `id`, stale bucket repopulation after `clear()` not resurrecting rows, legacy inline-state markers, non-cache facades, and non-map rows. That is a reasonable superset of the Python/PHP native suites, so the Rust port is no longer the untested language here.
- `arrayCacheShadow.py` is genuinely offline by default (`asyncio.run(offline())` always, `live()` only under `--live`), never loads keys, and its offline mode self-tests the comparator by corrupting a baseline row and asserting the mismatch is detected — good, that avoids a silently-vacuous differential.
- The transpiler hook is gated on `names.includes('test.cache')`, so it cannot leak into unrelated generated aggregators.

Two portability notes inline; neither blocks the perf work.

Nothing in this commit changes the behaviour I signed off on at `47f99f4`, and it strengthens the evidence base rather than the claims.

Merge gate: 🟢

**Comment by pcriadoperez** [comment]
Follow-up completed and pushed as b53fad30b5a (supersedes my earlier blocked-status comment).

- Required pre-push checks pass. Upstream lint/Rust-runner blockers resolved; full JS/Python/PHP/C#/Go/Rust WS base suites pass, and all 37 Java cache tests pass.
- Public Kraken same-input Python shadow passed for 180 seconds in each newUpdates mode: 8,709 watchOrderBook results, 193 watchTrades results, 198 watchOHLCV results, 121 large-slice checks, zero mismatches. No keys/private endpoints/mutating calls; connections closed.
- Coverage: 100% measured cache lines/statements in JS/Python/PHP/C#/Go/Java; Rust factories 100%, actual cache helper ranges 400/412 lines (97.1%). Remaining Rust lines and unmeasured PHP branches are explicitly documented, not hidden by exclusions.
- Runnable native/live tests and coverage notes are in build/tests/README.md. Python serialized-use and shared-row limitations still apply; Go locking remains separate in #30314.

Fresh remote CI is still running; this is not a claim that the full external live matrix is green. PR remains draft.
